test/infrastructure/docker/controllers/hollowmachine_controller.go

```diff
@@ -0,0 +1,221 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha4"
+	"sigs.k8s.io/cluster-api/controllers/remote"
+	infrav1 "sigs.k8s.io/cluster-api/test/infrastructure/docker/api/v1alpha4"
+	"sigs.k8s.io/cluster-api/test/infrastructure/docker/kubemark"
+	"sigs.k8s.io/cluster-api/util"
+	"sigs.k8s.io/cluster-api/util/annotations"
+	"sigs.k8s.io/cluster-api/util/patch"
+	"sigs.k8s.io/cluster-api/util/predicates"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// HollowMachineReconciler reconciles a HollowMachine object.
+type HollowMachineReconciler struct {
+	client.Client
+	Tracker *remote.ClusterCacheTracker
+}
+
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=hollowmachines,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=hollowmachines/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusters;machines,verbs=get;list;watch
+// +kubebuilder:rbac:groups="",resources=secrets;,verbs=get;list;watch
+
+// Reconcile handles HollowMachine events.
+func (r *HollowMachineReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, rerr error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	// Fetch the HollowMachine instance.
+	hollowMachine := &infrav1.HollowMachine{}
+	if err := r.Client.Get(ctx, req.NamespacedName, hollowMachine); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	// Fetch the HollowKubelet.
+	machine, err := util.GetOwnerMachine(ctx, r.Client, hollowMachine.ObjectMeta)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if machine == nil {
+		log.V(2).Info("Waiting for Machine Controller to set OwnerRef on the HollowMachine")
+		return ctrl.Result{}, nil
+	}
+
+	log = log.WithValues("machine", machine.Name)
+
+	if util.IsControlPlaneMachine(machine) {
+		return ctrl.Result{}, errors.New("HollowMachine can't be used for control plane machines")
+	}
+
+	// Fetch the Cluster.
+	cluster, err := util.GetClusterFromMetadata(ctx, r.Client, machine.ObjectMeta)
+	if err != nil {
+		log.Info("HollowMachine owner HollowKubelet is missing cluster label or cluster does not exist")
```

Comment:
should this log as error?

Reply:
we are alrading returning the error , which cause log.Error + exponential backoff